Libraries/Network/RCTNetworking.m

```diff
@@ -389,17 +402,44 @@ - (void)sendRequest:(NSURLRequest *)request
     });
   };
 
-  void (^incrementalDataBlock)(NSData *) = incrementalUpdates ? ^(NSData *data) {
-    dispatch_async(_methodQueue, ^{
-      [self sendData:data forTask:task];
-    });
-  } : nil;
+  // XHR does not allow you to peek at xhr.response before the response is
+  // finished. Only when xhr.responseType is set to ''/'text', consumers may
+  // peek at xhr.responseText. So unless the requested responseType is 'text',
+  // we only send progress updates and not incremental data updates to JS here.
+  RCTURLRequestIncrementalDataBlock incrementalDataBlock = nil;
+  RCTURLRequestProgressBlock downloadProgressBlock = nil;
+  if (incrementalUpdates) {
+    if ([responseType isEqualToString:@"text"]) {
+      incrementalDataBlock = ^(NSData *data, int64_t progress, int64_t total) {
+        dispatch_async(_methodQueue, ^{
+          NSString *responseString = [self decodeTextData:data fromResponse:task.response];
+          if (!responseString) {
+            RCTLogWarn(@"Received data was not a string, or was not a recognised encoding.");
+            return;
+          }
+          NSArray<id> *responseJSON = @[task.requestID, responseString, @(progress), @(total)];
+          [self sendEventWithName:@"didReceiveNetworkIncrementalData" body:responseJSON];
+        });
+      };
+    } else {
+      downloadProgressBlock = ^(int64_t progress, int64_t total) {
+        dispatch_async(_methodQueue, ^{
+          NSArray<id> *responseJSON = @[task.requestID, @(progress), @(total)];
+          [self sendEventWithName:@"didReceiveNetworkDataProgress" body:responseJSON];
+        });
+      };
+    }
+  }
 
   RCTURLRequestCompletionBlock completionBlock =
   ^(NSURLResponse *response, NSData *data, NSError *error) {
     dispatch_async(_methodQueue, ^{
-      if (!incrementalUpdates) {
-        [self sendData:data forTask:task];
+      // Unless we were sending incremental (text) chunks to JS, all along, now
+      // is the time to send the request body to JS.
+      if (!([responseType isEqualToString:@"text"] && incrementalUpdates)) {
```

Comment:
Can you turn the order of the condition around here, so it mirrors what you do on line 411.